File: SLAM_zero_to_hero/1_16/ros1/src/slam_tutorial/scripts/publisher.py
#!/usr/bin/env python3
"""
ROS1 Publisher Example

Publishes robot pose messages at 10 Hz.
Demonstrates basic ROS1 Python publisher pattern.
"""

import rospy
from geometry_msgs.msg import PoseStamped
import math


def main():
    # Initialize node
    rospy.init_node('pose_publisher', anonymous=True)

    # Create publisher
    pub = rospy.Publisher('/robot/pose', PoseStamped, queue_size=10)

    # Set rate (10 Hz)
    rate = rospy.Rate(10)

    rospy.loginfo("Pose publisher started")

    # Simulation time
    t = 0.0

    while not rospy.is_shutdown():
        # Create message
        msg = PoseStamped()
        msg.header.stamp = rospy.Time.now()
        msg.header.frame_id = "map"

        # Simulate circular motion
        msg.pose.position.x = math.cos(t * 0.1) * 5.0
        msg.pose.position.y = math.sin(t * 0.1) * 5.0
        msg.pose.position.z = 0.0

        # Simple quaternion (yaw only)
        yaw = t * 0.1
        msg.pose.orientation.x = 0.0
        msg.pose.orientation.y = 0.0
        msg.pose.orientation.z = math.sin(yaw / 2)
        msg.pose.orientation.w = math.cos(yaw / 2)

        # Publish
        pub.publish(msg)

        if int(t) % 10 == 0:
            rospy.loginfo(f"Published pose: x={msg.pose.position.x:.2f}, y={msg.pose.position.y:.2f}")

        t += 0.1
        rate.sleep()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: SLAM_zero_to_hero/part1_ch02_09/ros1/src/slam_tutorial/src/publisher.cpp
/**
 * ROS1 C++ Publisher Example
 *
 * Publishes robot pose messages at 10 Hz.
 */

#include <ros/ros.h>
#include <geometry_msgs/PoseStamped.h>
#include <cmath>

int main(int argc, char** argv) {
    // Initialize ROS
    ros::init(argc, argv, "pose_publisher_cpp");
    ros::NodeHandle nh;

    // Create publisher
    ros::Publisher pub = nh.advertise<geometry_msgs::PoseStamped>("/robot/pose", 10);

    // Set rate (10 Hz)
    ros::Rate rate(10);

    ROS_INFO("C++ Pose publisher started");

    double t = 0.0;

    while (ros::ok()) {
        // Create message
        geometry_msgs::PoseStamped msg;
        msg.header.stamp = ros::Time::now();
        msg.header.frame_id = "map";

        // Simulate circular motion
        msg.pose.position.x = std::cos(t * 0.1) * 5.0;
        msg.pose.position.y = std::sin(t * 0.1) * 5.0;
        msg.pose.position.z = 0.0;

        // Simple quaternion (yaw only)
        double yaw = t * 0.1;
        msg.pose.orientation.x = 0.0;
        msg.pose.orientation.y = 0.0;
        msg.pose.orientation.z = std::sin(yaw / 2);
        msg.pose.orientation.w = std::cos(yaw / 2);

        // Publish
        pub.publish(msg);

        if (static_cast<int>(t) % 10 == 0) {
            ROS_INFO("Published pose: x=%.2f, y=%.2f",
                     msg.pose.position.x, msg.pose.position.y);
        }

        t += 0.1;
        ros::spinOnce();
        rate.sleep();
    }

    return 0;
}
